### PR TITLE
BUGFIX: Simplify PartialPublish & Discard (for now) to NOT select dimensions and have custom events

### LINKED ISSUE #4614: BUG: Cut & paste in different dimension leads to unpublishable change
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When I copy a page to another dimension with copy & create, moving a content element with cut & past will move the element in all dimensions. 
Then a non-publishable change appears in the workspace that cannot be applied with 'publish', 'publish all' or using the workspace module. 

### Expected Behavior

I would expect to be able to publish/unpublish the change with 'publish all' or discard the change with 'discard changes' : 
Bildschirmfoto 2023-10-13 um 13 27 23

### Steps To Reproduce

Using Neos Demo: 

1. Copy 'Features' -> 'Multiple columns' to german using the 'create & copy' feature 
2. Publish everything 
3. Copy 'Features' -> 'Text & images' to german using the 'create & copy' feature
4. Publish everything
5. Go to page 'Multiple columns' in german & cut the heading 'A Caucus-race and a Long Tale' 
6. Go to page 'Text & images' and paste the heading 

Result: the element is moved in english and german and an unpublishable change appears 

### Environment

```markdown
- Flow:
- Neos: 9.0
- PHP: 8.2
```

### Anything else?

_No response_

### LINKED ISSUE #5387: PartialPublish / Discard vs Content Dimensions
_The versus issues goes into a next round. This time featuring our favourites: The opponent is publishing and fighting against content dimensions that are known to be especially cruel to ensure your mind does not sustain their presence._

--------

In Neos 8.3 we tested the following behaviour (via the Neos Ui / Workspace module):

1.) Publish All publishes without respecting the current dimension all changes in the site
2.) Publish Current (the document) only publishes the document in the currently selected dimension
3.) Publish Document in the workspace module publishes all dimension for this document (EDIT: with https://github.com/neos/neos-development-collection/pull/3986 and Neos 7.3 its was actually possible to publish each dimension too)
4.) When creating a page and then translating its possible to publish only the translation event though the original creation is not published

---------

The current state of Neos 9 claims to support at its core dimension selective partial publishing but currently its not used by the Neos Ui. Reason for this is a regression when refactoring change evaluation in the Neos Ui in Neos 9: https://github.com/neos/neos-ui/pull/3752#discussion_r1864538328
Well its not totally correct that we dont do dimension selective publishing at all as we still evaluate the changes from the change projection and then just use all changes and all dimensions ... well not without bugs: https://github.com/neos/neos-development-collection/issues/4614#issuecomment-2470222193

There are two ways from here: Either fix dimension based publishing or fully remove it for the release.

As we do not support 4. at the core - we cannot do a `CreateNodeVariant` before `CreateNodeAggregateWIthNode` is issued, and resorting and transforming the commands would be way to crazy - the viable option would be to not support it.

Change to simplify `PublishIndividualNodesFromWorkspace` to only allow to select node aggregate ids but not their dimension: https://github.com/neos/neos-development-collection/pull/5385

Further not allowing dimension based publishing allows us to ignore a lot of edge cases for now:
- in which dimension did a `NodeMove` happen and do we want to publish across all dimensions or split up a gather all command?
- node type change can only be published across all dimensions
- aggregate scoped properties can only be published across all dimensions
- node creation and their variant can only be published across all dimensions (or creation first)

Fully removing is of course a behaviour change. So in order to allow us to restore the old behaviour at some point if we find a good concept, publishing in the Neos Ui dialog must look different and contain a warning wich other dimensions are affected! In the simple case of only editing text on the page it will not contain that warning and it feels like Neos 8.3, also moving content elements via scatter will not raise that warning as only the one dimension will be affected. -> so we need to mark all. changes in all dimensions as orange via change projection and evaluate the data from the change projection to display the warning accordingly

### PR DESCRIPTION
**Pull request by mhsdesign**
Partially fixes: #4614 (as we would publish the node move that is marked as moved in some random dimension)
Partially fixes: https://github.com/neos/neos-development-collection/issues/5387
See also slack: https://neos-project.slack.com/archives/C04PYL8H3/p1732899328791309

### Why the `NodeIdentifiersToPublishOrDiscard` is removed from the event

Only with https://github.com/neos/neos-development-collection/commit/f2c66e08c92be23ad0b860547c5d3c1b46210a66
the `NodeIdentifiersToPublishOrDiscard` was introduced to the `WorkspaceWasPartiallyPublished` and `DiscardIndividualNodesFromWorkspace` events. Previously they did not contain the information which "filter" was used to create the events. And strictly speaking - for the ESCR core content graph this information on the event is purely cosmetic.

The only use currently is the asset usage catchup hook that uses the node ids and their dimension if a workspace was discarded to remove usage data:

https://github.com/neos/neos-development-collection/blob/be1f1fe6f430ae2d44765c64d3492fe7f2dddfb4/Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php#L62

One could also optimise the content cache flusher for the partial discard case to not flush the whole workspace but only partial entries:

https://github.com/neos/neos-development-collection/pull/5160/files#diff-008f3460b749dbf869e63ba0602a1501990fa09cc37cd7fdaca6afa3789017e4R143

As we currently flush the whole workspace on discard:

https://github.com/neos/neos-development-collection/blob/902e08d249c7a5145b1b9522efc6d0f86d072cdc/Neos.Neos/Classes/Fusion/Cache/GraphProjectorCatchUpHookForCacheFlushing.php#L196

But this is not worth to optimise as we discussed: https://github.com/neos/neos-development-collection/issues/5175#issuecomment-2386698992

So as this is the only - discussable use-case - we dont want to make the event creation more complicated. Instead of gathering the information which nodes were really affected and which dimensions we just fully leave out this information in the event. (Like we do also with other events where we don't specify which nodes were all actually removed in NodeAggregateWasRemoved but rely on that the projection does what it needs).

First i tried to extract the information for the `$publishedNodes` and `$discardedNodes` as shown in this pr in `extractNodeAggregateIdsAndAffectedDimensionSpacePoints` https://github.com/neos/neos-development-collection/pull/5386 but its problems are further described there.

Further treating a partial and full discard the same also simplifies the catchup hooks.

### Why the distinction between full and partial publish is removed from the event

Without the need to specify the `NodeIdentifiersToPublishOrDiscard` (see above) there is no reason to differentiate between the events at all.

Thus we upcast the the partial publish and discard events to a publish or discard inside `fromArray`:

```php
public static function fromArray(array $values): EventInterface
{
    return new WorkspaceWasPublished(
        WorkspaceName::fromString($values['sourceWorkspaceName']),
        WorkspaceName::fromString($values['targetWorkspaceName']),
        ContentStreamId::fromString($values['newSourceContentStreamId']),
        ContentStreamId::fromString($values['previousSourceContentStreamId']),
    );
}
```

Both events share the same properties their counter parts have.
Well keep the `$publishedNodes` and `$discardedNodes` fields in the database as they don't do harm.

`WorkspaceWasPartiallyPublished` -> `WorkspaceWasPublished`

- `WorkspaceName $sourceWorkspaceName`
- `WorkspaceName $targetWorkspaceName`
- `ContentStreamId $newSourceContentStreamId`
- `ContentStreamId $previousSourceContentStreamId`
- ignore: `NodeIdsToPublishOrDiscard $publishedNodes`

`WorkspaceWasPartiallyDiscarded` -> `WorkspaceWasDiscarded`

- `WorkspaceName $workspaceName`
- `ContentStreamId $newContentStreamId`
- `ContentStreamId $previousContentStreamId`
- ignore: `NodeIdsToPublishOrDiscard $discardedNodes`

To allow to distinct what events are part of a (partial) publish or not, instead the causation id will be used in the future: https://github.com/neos/neos-development-collection/issues/3887 (to be implemented in a followup)

**Upgrade instructions**

TODO:
- [x] adjust tests that test that partial publish is a full publish
- [x] check against left todos!

**Review instructions**

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Comment by kitsunet** [comment]
Still don't like it, but probably best way forward for now.

**Comment by mhsdesign** [comment]
Soooo i stripped this pr a little futher to its needed parts. And its beautiful. +130 lines -400lines. With the last controversial commit another -200 lines.
So a lot complexity is gone ^^ the parts that were perviously in here - extracting the node ids and their affected dsp as metadata for the event - are now part of https://github.com/neos/neos-development-collection/pull/5386
This 'feature' is unused (expect on place in the asset usage) and fully untested in the core. 

Adjustment in the asset usage:

> [TASK: Simplify asset usage catchup hook to not optimise partial discard](https://github.com/neos/neos-development-collection/pull/5385/commits/e4ad10b4484bd9d8357b2c17e33a25538778a365)
> Instead of fine granular updating only the usage references on the nodes that were discarded, well throw away everything and then rely on the reapplied events what is kept.
>
> This is more consistent to how other parts react to publishing like the graph just forks a new content stream and doesnt care about the old data, the change projection behaves similar to the graph.
> And for the content cache flusher we had the discussion to use `discardedNodes` but opted against it and do a full flush there as well: https://github.com/neos/neos-development-collection/issues/5175#issuecomment-2386698992
> 
> This change allows us to get rid of the `discardedNodes` property (with dimensions) in the `WorkspaceWasPartiallyDiscarded` event.

File: Neos.ContentRepository.Core/Classes/Feature/RebaseableCommands.php [hunk 1]
```diff
@@ -67,11 +72,37 @@ public function separateMatchingAndRemainingCommands(
...
-        NodeIdsToPublishOrDiscard $nodeIds,
+        RebasableToOtherWorkspaceInterface $command,
+        NodeAggregateIds $nodeIdsToMatch,
     ): bool {
-        foreach ($nodeIds as $nodeId) {
-            if ($command->matchesNodeId($nodeId)) {
+        foreach ($nodeIdsToMatch as $nodeId) {
+            /**
+             * This match must contain all commands which are working with individual nodes, such that they are
+             * filterable whether they are applying their action to a NodeIdToPublish.
```
Comment (dlubitz) [inline Neos.ContentRepository.Core/Classes/Feature/RebaseableCommands.php]: I don't get what this means. What does "whether they are applying their action to a NodeIdToPublish" mean? AFAIS we have no "NodeIdToPublish" anymore.

**Comment by dlubitz** [comment]
I don't understand, why we need to remove the "WasPartially*" Events. This is exactly what we shouldn't do in a event sourced system. I see the point, that we don't need to handle that anymore, but this shouldn't be the way to go.

**Comment by mhsdesign** [comment]
Its not the classy thing but were beyond that - especially as its beta id say. The point why i exactly removed it are:

- a replay is NOT necessary and the migration is only needed for if you want to replay some time - so the current state of the db is fine
- at some point we _might_ reintroduce the WasPartially events with proper metdata and information, and to avoid having WasPartially_2 then its just easier to get rid of them now as we dont need them
- its easy to forget implementing one of the events like PublishWorkspace or the partial thing, to make it simpler to write catchup hooks its just sane to have one event

i also discussed this approach with christian and Sebastian and thing its the right one? - also in terms of reducing complexity for once.

**Review by kitsunet — APPROVED** [review]
Yeah as I said before I am not happy at all with this, but it's probably safer for now. The "lost" information is minimal, I don't see a scenario in which we retroactively would want, need to do something based on the partially (published|discarded) events.

File: Neos.ContentRepository.Core/Classes/Feature/RebaseableCommands.php [hunk 2]
```diff
@@ -67,11 +72,37 @@ public function separateMatchingAndRemainingCommands(
...
+                CreateNodeVariant::class,
+                TagSubtree::class,
+                UntagSubtree::class
+                    => $command->nodeAggregateId->equals($nodeId),
+                CopyNodesRecursively::class => $command->nodeAggregateIdMapping->getNewNodeAggregateId(
+                    $command->nodeTreeToInsert->nodeAggregateId
+                )?->equals($nodeId),
+                SetSerializedNodeReferences::class => $command->sourceNodeAggregateId->equals($nodeId),
+                // todo return false instead to allow change to be still kept?
+                default => throw new \RuntimeException(sprintf('Command %s cannot be matched against a node aggregate id. Partial publish not possible.', $command::class), 1645393655)
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/RebaseableCommands.php]: Todo - needs test - there is a bug in partial publish and discard which means that if there is a command on the workspace NOT in this list, no partial operation is allowed. This is a bit to hard. Instead simply keeping that change as remainder makes more sense imo?

**Comment by mhsdesign** [comment]
Okay i give in. @dlubitz youre right we dont need to rewrite history if we dont need to. i relaxed the `fromArray` to allow to return other objects than `self`. Now its happening at runtime.

- [TASK: Remove migration for partially publish events and handle them a…](https://github.com/neos/neos-development-collection/pull/5385/commits/3412594387fa436d108affa3ed960db832964680) 
- [TASK: Simplify upcasting of legacy events to be in fromArray instea…](https://github.com/neos/neos-development-collection/pull/5385/commits/135dc67e89bdf90f6cc88c71b76de34001ab461d)

Also the content stream pruner will still select the legacy events in the query having in mind, that they will be upcasted.

**Review by dlubitz — APPROVED** [review]

**Comment by bwaidelich** [comment]
I'm late to the party and only skimmed through the code so far.
I fail to understand why those `Partially` events ought to be removed.
IMO there's a crucial domain-specific difference between a full and a partial publish/discard. And this difference should not only be reflected by some metadata in the events.
But most probably I misunderstood things

**Comment by mhsdesign** [comment]
As discussed today, i added a `partial` flag to the events as we want keep the main idea to reduce the complexity (only one public event)

See: [TASK: Introduce "partial" flag on publish events](https://github.com/neos/neos-development-collection/pull/5385/commits/6bc3c55654b87c9d126f1e70a83bbbb00bdb6c59)

File: Neos.ContentRepository.Core/Classes/Feature/RebaseableCommands.php (hunk 2 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/RebaseableCommands.php]: Solved via https://github.com/neos/neos-development-collection/pull/5385/commits/fa58f0c0f2f59cc41a1864ba6808e126f2677fd5

**Comment by kitsunet** [comment]
I guess we can merge this then?

**Pull request merged**